decentralized-message-queue/src/DMQ/Protocol/LocalMsgNotification/Type.hs

```diff
@@ -0,0 +1,181 @@
+
+    -- | Reply with a list of messages, along with a flag indicating
+    -- whether the server has more messages that it can provide.
+    --
+    -- TODO: maybe this should just be always a blocking request?
+    --
+    MsgReply
+      :: forall blocking msg.
+         BlockingReplyList blocking msg
+      -> HasMore
+      -> Message (LocalMsgNotification msg) (StBusy blocking) StIdle
+
+    -- | Termination message, initiated by the server when the client is
+    -- making a blocking call for more messages.
+    --
+    -- TODO: is this message desireable/necessary?
+    --
+    MsgServerDone
+      :: Message (LocalMsgNotification msg) (StBusy StBlocking) StDone
+
+    -- | The client can terminate the exchange when it has the agency
+    --
+    MsgClientDone
+      :: Message (LocalMsgNotification msg) StIdle StDone
+
+  type StateAgency  (StBusy blocking)  = ServerAgency
+  type StateAgency  StIdle  = ClientAgency
+  type StateAgency  StDone  = NobodyAgency
+
+  type StateToken = SingMsgNotification
+
+instance ( NFData msg
+         ) => NFData (Message (LocalMsgNotification msg) from to) where
+  rnf (MsgRequest block)  = rnf block
+  rnf (MsgReply a !_more) = rnf a
+  rnf MsgServerDone{}     = ()
+  rnf MsgClientDone{}     = ()
+
+deriving instance (Eq msg) =>
+                  Eq (Message (LocalMsgNotification msg) from to)
+
+deriving instance (Show msg) =>
+                  Show (Message (LocalMsgNotification msg) from to)
+
+
+-- | The kind of the local message notification protocol, and the types of
+-- the states in the protocol state machine.
+--
+-- It is parameterised over the type of messages
+--
+type LocalMsgNotification :: Type -> Type
+type data LocalMsgNotification msg where
+  StIdle :: LocalMsgNotification msg
+  StBusy :: StBlockingStyle -> LocalMsgNotification msg
+  StDone :: LocalMsgNotification msg
+
+instance ( ShowProxy msg
+         ) => ShowProxy (LocalMsgNotification msg) where
+    showProxy _ =
+      "LocalMsgNotification " ++ showProxy (Proxy :: Proxy msg)
+
+
+-- | The kind of blocking state
+--
+type data StBlockingStyle where
+
+  -- | In this sub-state the reply need not be prompt, and must
+  -- contain non-trivial payload.
+  --
+  StBlocking    :: StBlockingStyle
+
+  -- | In this state the peer shall respond promptly, possibly
+  -- with trivial/empty payload.
+  --
+  StNonBlocking :: StBlockingStyle
+
+
+-- | The value level equivalent of 'BlockingStyle'.
+--
+-- This is also used in 'MsgRequest' where it is interpreted (and can be
+-- encoded) as a 'Bool' with 'True' for blocking, and 'False' for non-blocking.
+--
+data SingBlockingStyle (k :: StBlockingStyle) where
+  SingBlocking    :: SingBlockingStyle StBlocking
+  SingNonBlocking :: SingBlockingStyle StNonBlocking
+
+deriving instance Eq   (SingBlockingStyle b)
+deriving instance Show (SingBlockingStyle b)
+
+instance NFData (SingBlockingStyle b) where
+  rnf SingBlocking    = ()
+  rnf SingNonBlocking = ()
```

Comment:
We don't need `NFData` instances for singletons, `!` already evaluates them to normal form :smile:.